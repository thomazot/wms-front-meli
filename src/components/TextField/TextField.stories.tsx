import { Story, Meta } from '@storybook/react'
import TextField, { TextFieldProps } from './TextField'

export default {
  title: 'Form/TextField',
  component: TextField,
  args: {
    label: 'E-mail',
    name: 'email',
    initialValue: '',
    placeholder: 'exemplo@gmail.com',
    disabled: false
  },
  argTypes: {
    onInput: { action: 'changed' }
  }
} as Meta

export const Default: Story<TextFieldProps> = (args) => (
  <div style={{ maxWidth: 300, padding: 15 }}>
    <TextField {...args} />
  </div>
)

export const withError: Story<TextFieldProps> = (args) => (
  <div style={{ maxWidth: 300, padding: 15 }}>
    <TextField {...args} />
  </div>
)

withError.args = {
  error: 'Ops...something is wrong'
}
