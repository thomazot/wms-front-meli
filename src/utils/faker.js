import faker from 'faker'

export const generator = (schema, min = 1, max) => {
  max = max || min
  return Array.from({
    length: faker.datatype.number({ min: min, max: max })
  }).map(() =>
    Object.keys(schema).reduce((entity, key) => {
      entity[key] = String(schema[key]).match(/^{{/)
        ? faker.fake(schema[key])
        : schema[key]
      return entity
    }, {})
  )
}

const productsSchema = {
  id: '{{datatype.uuid}}',
  title: '{{company.companyName}} {{company.companySuffix}}',
  price: '{{commerce.price}}',
  thumbnail: '/images/img.png',
  permalink: 'https://www.mercadolivre.com.br/produto/123',
  rating: 4,
  currency_id: 'BRL'
}

export const productsData = generator(productsSchema, 9, 9)

const menuSchema = {
  id: '{{datatype.uuid}}',
  name: '{{company.companyName}} {{company.companySuffix}}'
}

export const menuData = generator(menuSchema, 10, 20)
