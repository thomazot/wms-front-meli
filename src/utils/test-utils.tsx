import { ReactElement } from 'react'
import { render, RenderOptions } from '@testing-library/react'

import { ThemeProvider } from 'styled-components'
import theme from 'styles/theme'

import { WishlistProvider } from 'hooks/use-wishlist'

type CustomRenderProps = Omit<RenderOptions, 'queries'>

const customRender = (
  ui: ReactElement,
  { ...renderOptions }: CustomRenderProps = {}
) =>
  render(
    <WishlistProvider>
      <ThemeProvider theme={theme}>{ui}</ThemeProvider>
    </WishlistProvider>,
    renderOptions
  )

export * from '@testing-library/react'
export { customRender as render }
