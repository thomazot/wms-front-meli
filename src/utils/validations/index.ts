import Joi from 'joi'

const fieldsValidations = {
  email: Joi.string()
    .email({ tlds: { allow: false } })
    .required(),
  password: Joi.string().required()
}

function getFieldErros(objError: Joi.ValidationResult) {
  const errors: FieldErrors = {}

  if (objError.error) {
    objError.error.details.forEach((detail) => {
      errors[detail.path.join('.')] = detail.message
    })
  }

  return errors
}

export type FieldErrors = {
  [key: string]: string
}

export type SiginInValues = {
  email: string
  password: string
}

export function signInValidate(values: SiginInValues): FieldErrors {
  const { email, password } = fieldsValidations
  const schema = Joi.object({ email, password })

  return getFieldErros(schema.validate(values, { abortEarly: false }))
}
