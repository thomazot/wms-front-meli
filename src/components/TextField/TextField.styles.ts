import styled, { css, DefaultTheme } from 'styled-components'
import { TextFieldProps } from './TextField'

type WrapperProps = Pick<TextFieldProps, 'disabled'> & { error?: boolean }

export const InputWrapper = styled.div`
  ${({ theme }) => css`
    display: flex;
    align-items: center;
    border-radius: ${theme.border.radius};
    padding: 0 ${theme.spacings.xsmall};
    border: 0.2rem solid ${theme.colors.primary.lightest};
    &:focus-within {
      box-shadow: 0 0 0.5rem ${theme.colors.primary.lightest};
    }
  `}
`

export const Label = styled.label`
  ${({ theme }) => css`
    display: block;
    font-size: ${theme.font.sizes.small};
    color: ${theme.colors.base};
    cursor: pointer;
    padding: ${theme.spacings.xxsmall} 0;
  `}
`

export const Input = styled.input`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.medium};
    padding: ${theme.spacings.xxsmall} 0;
    background: transparent;
    border: 0;
    outline: none;
    width: 100%;
    &:-webkit-autofill {
      filter: none;
      &::first-line {
        font-size: ${theme.font.sizes.medium};
      }
    }
    &:focus::placeholder {
      color: transparent;
    }
  `}
`

export const Error = styled.div`
  ${({ theme }) => css`
    color: ${theme.colors.feedback.error.action};
    font-size: ${theme.font.sizes.xsmall};
    padding: ${theme.spacings.xxsmall} 0;
  `}
`

const wrapperModifiers = {
  error: (theme: DefaultTheme) => css`
    ${InputWrapper} {
      border-color: ${theme.colors.feedback.error.active};
    }
    ${Label} {
      color: ${theme.colors.feedback.error.action};
    }
  `,
  disabled: () => css`
    ${Label},
    ${Input} {
      cursor: not-allowed;
      &::placeholder {
        color: currentColor;
      }
    }
  `
}

export const Wrapper = styled.div<WrapperProps>`
  ${({ theme, error, disabled }) => css`
    ${error && wrapperModifiers.error(theme)}
    ${disabled && wrapperModifiers.disabled()}
  `}
`
