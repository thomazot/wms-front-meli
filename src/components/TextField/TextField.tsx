import { useState, InputHTMLAttributes } from 'react'

import * as S from './TextField.styles'

export type TextFieldProps = {
  onInputChange?: (value: string) => void
  label?: string
  initialValue?: string
  disabled?: boolean
  error?: string
} & InputHTMLAttributes<HTMLInputElement>

const TextField = ({
  label,
  name,
  onInputChange,
  initialValue = '',
  disabled = false,
  error,
  ...props
}: TextFieldProps) => {
  const [value, setValue] = useState(initialValue)

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const currentValue = event.currentTarget.value

    setValue(currentValue)

    !!onInputChange && onInputChange(currentValue)
  }

  return (
    <S.Wrapper disabled={disabled} error={!!error}>
      {!!label && <S.Label htmlFor={name}>{label}</S.Label>}
      <S.InputWrapper>
        <S.Input
          onChange={handleChange}
          value={value}
          name={name}
          disabled={disabled}
          {...(label ? { id: name } : {})}
          {...props}
        />
      </S.InputWrapper>
      {!!error && <S.Error>{error}</S.Error>}
    </S.Wrapper>
  )
}
export default TextField
