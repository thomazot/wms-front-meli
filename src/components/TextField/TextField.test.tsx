import { render, screen, waitFor } from 'utils/test-utils'
import userEvent from '@testing-library/user-event'

import TextField from '.'

describe('<TextField />', () => {
  it('should render component', () => {
    const { container } = render(<TextField />)

    expect(container).toBeInTheDocument()
  })

  it('should render with label', () => {
    render(<TextField label="Label" name="Label" />)

    expect(screen.getByLabelText('Label')).toBeInTheDocument()
  })

  it('should without label', () => {
    render(<TextField />)

    expect(screen.queryByLabelText('Label')).not.toBeInTheDocument()
  })

  it('should render with placeholder', () => {
    render(<TextField placeholder="Placeholder" />)

    expect(screen.getByPlaceholderText('Placeholder')).toBeInTheDocument()
  })

  it('should change its value when typing', async () => {
    const onInputChange = jest.fn()
    render(<TextField name="field" onInputChange={onInputChange} />)

    const input = screen.getByRole('textbox')
    const text = 'This is my text'

    userEvent.type(input, text)

    await waitFor(() => {
      expect(input).toHaveValue(text)
      expect(onInputChange).toHaveBeenCalledTimes(text.length)
    })

    expect(onInputChange).toHaveBeenCalledWith('field', text)
  })

  it('should does not changes its value when disabled', async () => {
    const onInputChange = jest.fn()
    render(<TextField onInputChange={onInputChange} disabled />)

    const input = screen.getByRole('textbox')
    expect(input).toBeDisabled()

    const text = 'This is my text'
    userEvent.type(input, text)

    await waitFor(() => {
      expect(input).not.toHaveValue(text)
      expect(onInputChange).toHaveBeenCalledTimes(0)
    })
    expect(onInputChange).not.toHaveBeenCalled()
  })

  it('should renders with error', () => {
    const { container } = render(<TextField error="Error message" />)

    expect(screen.getByText('Error message')).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
  })
})
