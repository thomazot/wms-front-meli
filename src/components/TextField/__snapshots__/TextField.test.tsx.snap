// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> should renders with error 1`] = `
.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 0.4rem;
  padding: 0 1.6rem;
  border: 0.2rem solid #C5DCFF;
}

.c2:focus-within {
  box-shadow: 0 0 0.5rem #C5DCFF;
}

.c3 {
  font-size: 1.6rem;
  padding: 0.8rem 0;
  background: transparent;
  border: 0;
  outline: none;
  width: 100%;
}

.c3:-webkit-autofill {
  -webkit-filter: none;
  filter: none;
}

.c3:-webkit-autofill::first-line {
  font-size: 1.6rem;
}

.c3:focus::-webkit-input-placeholder {
  color: transparent;
}

.c3:focus::-moz-placeholder {
  color: transparent;
}

.c3:focus:-ms-input-placeholder {
  color: transparent;
}

.c3:focus::placeholder {
  color: transparent;
}

.c4 {
  color: #B33616;
  font-size: 1.2rem;
  padding: 0.8rem 0;
}

.c0 .c1 {
  border-color: #EFB5AC;
}

<div
  class="c0"
>
  <div
    class="c1 c2"
  >
    <input
      class="c3"
      value=""
    />
  </div>
  <div
    class="c4"
  >
    Error message
  </div>
</div>
`
